test/assert.js

```diff
@@ -648,6 +652,17 @@ test('.throws() fails if passed a bad value', t => {
 	t.end();
 });
 
+test('promise .throws() fails when promise is resolved', t => {
+	return assertions.throws(Promise.resolve('foo'))
+	.catch(err => {
+		testFailResults(t, err, {
```

Comment:
My bad, overlooked that `failsWith()` resets `lastFailure`.

Instead extract the failure testing code into its own function, e.g.:

```js
function assertFailure() {
  if (!lastFailure) {
  		t.fail('Expected assertion to fail');
 		return;
 	}
 
 	t.is(lastFailure.assertion, subset.assertion);
 	t.is(lastFailure.message, subset.message);
 	t.is(lastFailure.name, 'AssertionError');
 	t.is(lastFailure.operator, subset.operator);
 	if (subset.statements) {
 		t.is(lastFailure.statements.length, subset.statements.length);
 		lastFailure.statements.forEach((s, i) => {
 			t.is(s[0], subset.statements[i][0]);
 			t.match(s[1], subset.statements[i][1]);
 		});
 	} else {
 		t.same(lastFailure.statements, []);
 	}
 	if (subset.values) {
 		t.is(lastFailure.values.length, subset.values.length);
 		lastFailure.values.forEach((s, i) => {
 			t.is(s.label, subset.values[i].label);
 			t.match(s.formatted, subset.values[i].formatted);
 		});
 	} else {
 		t.same(lastFailure.values, []);
  	}
  }
}
```

And call that from `failsWith()` and `eventuallyFailsWith()` (so the latter doesn't call `failsWith()` anymore). `eventuallyFailsWith()` should also reset `lastFailure`.